R/add_cols.R

```diff
@@ -118,60 +118,97 @@ NULL
 }
 
 #' @name .add_date
-.add_deaths <- function(.data,
-                        onset_to_death,
-                        hosp_death_risk,
-                        non_hosp_death_risk) {
+.add_outcome <- function(.data,
+                         onset_to_death,
+                         onset_to_recovery,
+                         hosp_death_risk,
+                         non_hosp_death_risk,
+                         config) {
   infected_idx <- .data$infected == "infected"
   num_infected <- sum(infected_idx)
-  .data$deaths <- NA_real_
-  .data$deaths[infected_idx] <- .data$time[infected_idx] +
-    onset_to_death(num_infected)
+  .data$outcome <- "contact"
+  .data$outcome_time <- NA_real_
+  .data$outcome[infected_idx] <- "recovered"
+  .data$outcome_time[infected_idx] <- .data$time[infected_idx] +
+    onset_to_recovery(num_infected)
+  hosp_idx <- !is.na(.data$hospitalisation) & infected_idx
+  non_hosp_idx <- is.na(.data$hospitalisation) & infected_idx
 
-  apply_death_risk <- function(.data, risk, hosp = TRUE) {
+  # define functions used only by .add_outcome
+  time_varying_risk <- function(.data, died_idx, config) {
+    # check time-varying function
+    .check_func_req_args(config$time_varying_death_risk)
+
+    # sample which deaths to keep with time-varying risk
+    prob_death <- config$time_varying_death_risk(.data$time[died_idx])
+    max_prob_death <- stats::optimise(
+      f = config$time_varying_death_risk,
+      interval = c(0, 100),
+      maximum = TRUE
+    )$objective
+    norm_prob_death <- prob_death / max_prob_death
+    # set all values > 1 and < 0 to 1 and 0 due to imprecision of optimise
+    # TODO: ideally this step would be removed as no values are 0 > x < 1
+    norm_prob_death[norm_prob_death > 1] <- 1
+    norm_prob_death[norm_prob_death < 0] <- 0
```

Comment:
This was added because the optimisation is not exact and so the maximum value of the time-varying cfr function provided by the user might be greater than the maximum found by `optimise()`, which leads the `prob_death / max_prob_death` to produce values greater than 1, which is not allowed under this implementation where the time-varying fatality risk scaling is between 1 (i.e. maximum fatality risk at time $t$), or 0 (i.e. minimum fatality risk at time $t$, which is equivalent to guaranteed recovery).

As the comment above states, ideally this would be removed, but I'm yet to think of how this can be done without breaking the code. By leaving the comment it reminds me and other developers to work on this in the future, but I don't think this should be a blocker for merging the feature. All cases of `norm_prob_death` being above 1 or below 0 should be due to numerical imprecision, so by setting them to 0 or 1 it does not drastically change the value.

Happy to hear thoughts on how this can be solved going forward, either in this PR or as an issue.

Reply:
I _think_ a simple rescaling function would suffice, like so `f = function(x, n = 0.9) (x / max(x)) * n` where `n` can be `max_prob_death`. This func assumes your lower bound is zero, which would be good to add as a check for the time-dependent CFR function `cofig$time_varying_death_risk()`.

Reply:
I'm not sure I understand your answer and proposed function, and which part of the existing code it is replacing. Could you provide a code chunk that applies the function you suggested to the current function (it doesn't have to be the whole function just 5 lines before and 5 lines after the proposed function call)?

Reply:
I'm also a bit confused now. My initial thought was that you want to scale the `prob_death` vector to the range `[0.0, max_prob_death]` - but that's not actually what's happening.

Here's a reprex that shows what I think the code is doing and my solution.

``` r
# a step-wise CFR, expected values are 0.001 or 0.005
f = function(x) {
  ifelse(test = x > 50, yes = 0.005, no = 0.001)
}

f(10)
#> [1] 0.001
f(60)
#> [1] 0.005

# your code to find the function maximum
max_prob = stats::optimise(
  f = f, interval = c(0, 100), maximum = TRUE
)$objective
max_prob
#> [1] 0.005

# normalising prob_death against max_prob
norm_prob_death = f(c(10, 20, 60)) / max_prob

# this leads to really high values of normalised prob death
# is this what is expected?
norm_prob_death
#> [1] 0.2 0.2 1.0

#### my solution ####
# assume max_prob is found to be 0.004 for some reason
max_prob2 = 0.004

# still very high normalised values, and some > 1.0
f(c(10, 20, 60)) / max_prob2
#> [1] 0.25 0.25 1.25

# rescale to the range [0, max_prob2]
norm_prob_death2 = scales::rescale(f(c(10, 20, 60)), to = c(0, max_prob2))

# this seems much more reasonable?
norm_prob_death2
#> [1] 0.000 0.000 0.004
```

<sup>Created on 2024-04-19 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

Reply:
The `prob_death` needs to be scaled to `[0, 1]` not `[0, max_death_rate]` because it is used in the binomial sampling later in the function.

`optimise()` is run over the interval `[0, 100]` because if the upper bound of the interval is higher it can fail to find the function maximum. If the function maximum is above 100 then this code will fail to find the maximum and return the local maximum between 0 and 100 and so the code will technically be wrong. I was meant to make a comment in the code or an issue to track this.

The problem is because the user can input any anonymous function it's impossible to know where to search, but `[0, 100]` seemed realistic and is narrow enough for the search to reliably find the optimum. 

Please can you log this as an issue and I will tackle it in another PR?

One option is to add another argument to `config` which is something along the lines of `time_varying_cfr_max_range` which by default could be `c(0, 100)`, but allows the user to change it.

Reply:
The code highlighted in this comment has been removed since a1fdbb5372961d0eb4c54c6794ec68cfb4321d31. The time-varying cfr has been simplified in the last few commits on this feature branch. 

These changes come from a discussion I had with @pratikunterwegs about simplifying the way time-varying death rate is calculated in the simulation. The new anonymous function signature required for the time-varying cfr is demonstrated in the `time-varying-cfr.Rmd` vignette.